File: src/main/java/com/coralblocks/coralbench/bench/LongMapBench.java
package com.coralblocks.coralbench.bench;

import java.text.DecimalFormat;

import com.coralblocks.coralbench.Bench;
import com.coralblocks.coralbench.util.LongMap;
import com.coralblocks.coralbench.util.Random;
import com.coralblocks.coralbench.util.ThreadPinning;

public class LongMapBench {
	
	private static final long THE_SEED = 123456789L;
	private static final DecimalFormat FORMATTER = new DecimalFormat("#,###");
	
	public static void main(String[] args) {
		
		final int warmupIterations = Integer.parseInt(args[0]);
		
		final int benchIterations = Integer.parseInt(args[1]);
		
		final int passes = Integer.parseInt(args[2]);
		
		final int coreId = Integer.parseInt(args[3]);
		
		if (coreId >= 0) ThreadPinning.pinCurrentThread(coreId);
		
		final String filler = new String("FILLER");
		
		LongMap<String> map = new LongMap<String>(8388608); // 2 ^ 23
		
		Random rand = new Random(THE_SEED);
		
		Bench bench = new Bench(warmupIterations);
		
		System.out.println();
		
		System.out.println("Benchmarking put operation...");
		
		System.out.println();
		
		for(int pass = 0; pass < passes; pass++) {

			if (pass > 0) { // no need to reset on the first pass...
				bench.reset(false);
				rand.reset();
			}
			
			// first pass has warmup... after that there is no warmup for subsequent passes...
			int totalIterations = pass == 0 ? (warmupIterations + benchIterations) : benchIterations;
			
			map.clear();
		
			long time = System.nanoTime();
			
			for(int i = 0; i < totalIterations; i++) {
				
				long key = rand.nextLong();
				
				bench.mark();
				map.put(key, filler);
				bench.measure();
			}
			
			time = System.nanoTime() - time;
			
			System.out.println("Total time in nanoseconds: " + FORMATTER.format(time) +  " (" + (pass == 0 ? "with" : "without") + " warmup)");
			System.out.println("Final size of map: " + map.size());
			System.out.println();
			bench.printResults();
		}
		
		System.out.println("Benchmarking get operation...");
		
		System.out.println();
		
		rand.reset(); // first pass below will not be reset so reset here...
		
		for(int pass = 0; pass < passes; pass++) {
			
			if (pass > 0) { // no need to reset on the first pass...
				bench.reset(false);
				rand.reset();
			}
			
			// first pass has warmup... after that there is no warmup for subsequent passes...
			int totalIterations = pass == 0 ? (warmupIterations + benchIterations) : benchIterations;
			
			String gotten = null;
		
			long time = System.nanoTime();
			
			for(int i = 0; i < totalIterations; i++) {
				
				long key = rand.nextLong();
				
				bench.mark();
				gotten = map.get(key);
				bench.measure();
			}
			
			time = System.nanoTime() - time;
			
			System.out.println("Total time in nanoseconds: " + FORMATTER.format(time) +  " (" + (pass == 0 ? "with" : "without") + " warmup)");
			System.out.println("Last object gotten: " + gotten);
			System.out.println();
			bench.printResults();
		}
	}
}

File: src/main/c/CoralBlocks/CoralBench/Bench/LongMapBench.cpp
#include <iostream>
#include <iomanip>
#include <ctime>
#include <cstdlib>
#include <unordered_map>

#include "../Bench.cpp"
#include "../Util/ThreadPinning.cpp"
#include "../Util/LongMap.h"
#include "../Util/Random.cpp"

using std::string;
using std::cout;
using std::cerr;
using std::endl;
using std::atoi;
using namespace CoralBlocks::CoralBench;
using namespace CoralBlocks::CoralBench::Util;

long get_nano_ts(struct timespec &ts) {
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec * 1000000000 + ts.tv_nsec;
}

string formatWithCommas(long value) {
    stringstream ss;
    ss.imbue(std::locale("en_US")); // for #,###
    ss << std::fixed << value;
    return ss.str();
}

int main(int argc, char* argv[]) {

    const long THE_SEED = 123456789L;

    if (argc < 4) {
        cerr << "Usage: " << argv[0] << " <iterations> <passes> <coreId>" << endl;
        return 1;
    }

    int iterations = atoi(argv[1]);
    int passes = atoi(argv[2]);
    int coreId = atoi(argv[3]);

    if (coreId >= 0) ThreadPinning::pinCurrentThread(coreId);

    LongMap<string>* mapPointer = new LongMap<string>(8388608);  // 2 ^ 23
    Random* randPointer = new Random(THE_SEED);
    Bench* benchPointer = new Bench();

    LongMap<string>& map = *mapPointer;
    Random& rand = *randPointer;
    Bench& bench = *benchPointer;

    struct timespec ts;
    string filler("FILLER");

    cout << "\nBenchmarking put operation...\n";

    for (int pass = 0; pass < passes; pass++) {

        map.clear();  // we will re-insert everything...
        bench.reset();
        rand.reset();

        long time = get_nano_ts(ts);

        for (int i = 0; i < iterations; i++) {
            long key = rand.nextLong();  // this is deterministic (pseudo-random)

            bench.mark();
            map.put(key, filler);
            bench.measure();
        }

        time = get_nano_ts(ts) - time;

        cout << "Total time in nanoseconds: " << formatWithCommas(time) << endl;
        cout << "Final size of map: " << map.size() << endl << endl;
        bench.printResults();
    }

    cout << "Benchmarking get operation...\n";

    for (int pass = 0; pass < passes; pass++) {
        
        bench.reset();
        rand.reset();

        string* gotten;

        long time = get_nano_ts(ts);

        for (int i = 0; i < iterations; i++) {

            long key = rand.nextLong();  // this is deterministic (pseudo-random)

            bench.mark();
            gotten = map.get(key);
            bench.measure();
        }

        time = get_nano_ts(ts) - time;

        cout << "Total time in nanoseconds: " << formatWithCommas(time) << endl;
        cout << "Last object gotten: " << *gotten << endl << endl;
        bench.printResults();
    }

    cout << "Benchmarking remove operation...\n";

    map.clear();

    for (int pass = 0; pass < passes; pass++) {

        rand.reset();
        for (int i = 0; i < iterations; i++) map.put(rand.nextLong(), filler);

        bench.reset();
        rand.reset();

        string* removed;

        long time = get_nano_ts(ts);

        for (int i = 0; i < iterations; i++) {

            long key = rand.nextLong();  // this is deterministic (pseudo-random)

            bench.mark();
            removed = map.remove(key);
            bench.measure();
        }

        time = get_nano_ts(ts) - time;

        cout << "Total time in nanoseconds: " << formatWithCommas(time) << endl;
        cout << "Last object gotten: " << *removed << endl;
        cout << "Final size of map: " << map.size() << endl << endl;
        bench.printResults();
    }

    delete mapPointer;
    delete benchPointer;
    delete randPointer;

    return 0;
}
